Add overloads for { parseAs: "string" } or { parseAs: "buffer" }

jineshshah36 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Tip: `npm run bench` to compare branches interactively.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

all tests passed

allevo commented on test/types/index.ts:
```diff
@@ -324,19 +324,23 @@ server.setSchemaCompiler(function (schema: object) {
 
 server.addSchema({})
 
-server.addContentTypeParser('foo/bar', {}, (req, done) => {
```
Could you keep this too?

jineshshah36 replied:
This conflicts with your docs. It says:
```
As you can see, now the function signature is (req, body, done) instead of (req, done)
```

allevo replied:
This is allowed. The body parameter is in the signature only if `returnAs` is specified

jineshshah36 replied:
https://github.com/fastify/fastify/pull/1162/files#r217922743

allevo commented on test/types/index.ts:
```diff
@@ -324,19 +324,23 @@ server.setSchemaCompiler(function (schema: object) {
   done!(null, {})
 })
 
-server.addContentTypeParser('foo/bar', { parseAs: 'string' }, (req, done) => {
```
This too

jineshshah36 replied:
This conflicts with your docs. It says:

```
As you can see, now the function signature is (req, body, done) instead of (req, done)
```

Additionally, if this is allowed, it seems like the body is getting parsed as a string and then ignored. What's the point?

allevo commented on test/types/index.ts:
```diff
@@ -324,19 +324,23 @@ server.setSchemaCompiler(function (schema: object) {
   done!(null, {})
 })
 
-server.addContentTypeParser('foo/bar', { bodyLimit: 20 }, (req, done) => {
```
And this!

jineshshah36 replied:
This conflicts with your docs. It says:
```
As you can see, now the function signature is (req, body, done) instead of (req, done)
```

jineshshah36 commented:
Wait I’m confused, is parseAs not required if 3 arguments are passed? If not, what will the body get parsed as? Also, if you do provide three arguments, doesn’t the callback require three arguments?

jineshshah36 commented:
I already read that but it is incomplete. For example, if I provide parseAs: 'string', but only 2 arguments in the callback, is it just swallowing the body? When does it even make sense to do this:

```
server.addContentTypeParser('foo/bar', { parseAs: 'string' }, (req, done) => {
  ...
})
```

allevo commented:
Maybe the documentation is not so clear.
Anyway reading the code

https://github.com/fastify/fastify/blob/d0d2aa9fe362d6164ff9945083a80494d9b8d434/lib/contentTypeParser.js#L68-L82

The signature is:
- `(req, done)` / `(req) => Promise` if the option is not given or if the `resultAs` is not specified
- `(req, body, done)` / `(req, body) => Promise` if and only if the `option.resultAs` is `string` or `buffer`

jineshshah36 commented:
Ok so based on that the following test would be invalid:

```
server.addContentTypeParser('foo/bar', { parseAs: 'string' }, (req, done) => {
    ...
})
```

since it should take 3 arguments in the callback.

I've updated the rest of the types and tests accordingly.

allevo commented on fastify.d.ts:
```diff
@@ -437,7 +442,10 @@ declare namespace fastify {
      * Add a content type parser
      */
-    addContentTypeParser(contentType: string, opts: {parseAs?: string, bodyLimit?: number} | AsyncContentTypeParser<HttpRequest> | ContentTypeParser<HttpRequest>, parser?: AsyncContentTypeParser<HttpRequest> | ContentTypeParser<HttpRequest>): void;
+    addContentTypeParser(contentType: string, opts: { bodyLimit?: number }, parser: ContentTypeParser<HttpRequest>): void
```
The array parameter is supported everywhere in `addContentTypeParser`

jineshshah36 replied:
Fixed

jineshshah36 commented:
Fixed the "The array parameter is supported everywhere in addContentTypeParser". And, I added back "server.addContentTypeParser('foo/bar', {}, (req, done) => {" already.

jineshshah36 commented on test/types/index.ts:
```diff
@@ -324,19 +324,31 @@ server.setSchemaCompiler(function (schema: object) {
 
-server.addContentTypeParser('foo/bar', { parseAs: 'string' }, (req, done) => {
-  done!(null, {})
+server.addContentTypeParser('foo/bar', {}, (req, done) => {
```
Here is where it is added back

jineshshah36 commented:
I would say that it's clear to use from the docs. It was complicated here because the type system needs to account for every possible valid way to use the api. Most of the calls in the tests are redundant ways to do the same thing. It's important that typescript types account for them, but I don't really think it matters if the docs provide them.

allevo commented:
The maintainers and the collaborators are not mainly typescript users. If you are interested, we'll be happy to review a PR about a rework of the typescript test suite!
